typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+  and probably exceptionally rare on the small systems Tock targets.
+* `char` is rare in Tock, as Tock mostly treats text as byte buffers. When
+  needed, it can be passed as `u32`. It's also not a standard part of C (where
+  `char` has a different meaning).
+* Wide pointers: these aren't a concept in other languages (e.g. C) and
+  therefore don't belong in a syscall ABI.
+* `*const T` and `*mut T` are redundant: they're distinct types, but only exist
+  to communicate mutability information. We already handle this in Tock by using
+  different system calls for buffers of different mutability. Therefore we only
+  need one pointer type.
+
+### The reduced list
+
+This leaves us with:
+
+* `bool`
+* Numeric: `i32`, `i64`, `isize`, `u32`, `u64`, `usize`, `f32`, `f64`
+* Upcall function pointer
+* Narrow pointer (`*const T` or `*mut T` where `T: Sized`)
+* `ErrorCode`
+* Non-pointer CHERI capability
+
+## Type descriptors
+
+When a process wants to send data to the kernel, the process needs a way to tell
+the kernel the sent data's type (and vice versa when the kernel sends data to
+the process). To do this, we need a way to serialize information about a a list
+of types. To start, lets assign numbers to each type (DNE means this type
+doesn't exist yet):
+
+| ID       | Description                  | Kernel Type     | `libtock-c` Type | `libtock-rs` Type         |
+| -------- | ---------------------------- | --------------- | ---------------- | ------------------------- |
+| `0b0001` | Error code                   | `ErrorCode`     | DNE              | `ErrorCode`               |
+| `0b0010` | `u32`                        | `u32`           | `uint32_t`       | `u32`                     |
+| `0b0011` | `i32`                        | `i32`           | `int32_t`        | `i32`                     |
+| `0b0100` | `usize`                      | `usize`         | `size_t`         | `usize`                   |
+| `0b0101` | `isize`                      | `isize`         | `ptrdiff_t`      | `isize`                   |
+| `0b0110` | `u64`                        | `u64`           | `uint64_t`       | `u64`                     |
+| `0b0111` | `i64`                        | `i64`           | `int64_t`        | `i64`                     |
+| `0b1000` | `f32`                        | `f32`           | `float`          | `f32`                     |
+| `0b1001` | `f64`                        | `f64`           | `double`         | `f64`                     |
+| `0b1010` | `bool`                       | `bool`          | `bool`           | `bool`                    |
+| `0b1011` | Upcall pointer               | `CapabilityPtr` | `UpcallFn`       | `UpcallFn`                |
+| `0b1100` | Data pointer                 | `CapabilityPtr` | `T*`             | `*mut T` where `T: Sized` |
+| `0b1101` | Non-pointer CHERI capability | DNE             | DNE              | DNE                       |
```

Comment:
Does this need to be fundamentally different from Arbitrary register value?

Really, I'm asking whether it needs to have a particular value reserved for it specially.

Reply:
Conceptually, I think it should be. For example, a filesystem that uses CHERI capabilities to manage file access would pass those capabilities as Command arguments and return values, and I think it's worth distinguishing that from "arbitrary register value" (which should probably be disallowed as a Command argument or return type).